components/brave_referrals/browser/brave_referrals_service.cc

```diff
@@ -252,6 +256,15 @@ void BraveReferralsService::OnReferralInitLoadComplete(
                << ", response code: " << response_code
                << ", payload: " << safe_response_body
                << ", url: " << referral_init_loader_->GetFinalURL().spec();
+    if (!installation_timer_) {
```

Comment:
I feel like it might be a little more concise to just create a repeating timer to initialize the referrals when the services starts, and then stop the timer when it succeeds (so usually the timer will just run a single time, except when the referrals server is unreachable).

Reply:
In my first iteration of the code I called `InitReferrals()` through a timer in `BraveReferralsService::OnReadPromoCodeComplete()`. It is simpler, but I thought about the edge case where the browser fires up, the timer gets scheduled 3 hours from now, and the user closes the browser after 30 minutes and never uses it again. We lost that data point as a valid installation. @aekeus what are your thoughts here?

Reply:
Oh yeah, we wouldn't want to do that. We've encountered this problem before in the referrals service. We usually run the required function once at startup and then periodically on the timer. See how we handle fetching the referral headers, for example.

Reply:
This code is basically doing what the referral headers does. The only difference is that request has a dependency on the promo code, which is read from a file, so it needs to be kicked off from another thread. The first run happens here. We also only want to hit this endpoint once.

https://github.com/brave/brave-core/pull/5144/commits/c8a70d22ad964ad94e6f14fdef8c91127b132b36#diff-73dd478ae750b6b4236d6f5de425367aL345

Reply:
OK, let's keep like it that then.